Enable saving and loading FP8 model (#1683)
huggingface/optimum-habana: examples/text-generation/README.md

When we run a model on single card and don't use deepspeed, the `--disk_offload` flag allows to offload weights to disk during model quantization in INC. When this flag is mentioned, during the quantization process, each weight first is loaded from disk to CPU RAM, when brought to HPU DRAM and quantized there. This way not all the model is on the CPU RAM but only one weight each time.
To enable this weights offload mechanism, add `--disk_offload` flag to the topology command line.
Here is an example of using disk_offload in quantize command.

> Please follow the [Running FP8 models on single device](#running-fp8-models-on-a-single-device) section first before running the cmd below.

Reviewer: ```suggestion
Please follow the [Running FP8 models on single device](#running-fp8-models-on-single-device) section first before running the cmd below.
```

--flash_attention_recompute
```

### Saving FP8 Checkpoints in Hugging Face format
After quantizing the model, we can save it to a local path.

> **Note:** Before executing the command below, please refer to the "Running FP8 Models on a Single Device" section to measure the model quantization statistics.

Reviewer: ```suggestion
> [!NOTE]  
> Before executing the command below, please refer to the "Running FP8 Models on a Single Device" section to measure the model quantization statistics.
```

--flash_attention_recompute
```

### Saving FP8 Checkpoints in Hugging Face format
After quantizing the model, we can save it to a local path.

> [!NOTE]  

> > Before executing the command below, please refer to the "Running FP8 Models on a Single Device" section to measure the model quantization statistics.

Reviewer: ```suggestion
> Before executing the command below, please refer to the ["Running FP8 Models on a Single Device" section](#running-fp8-models-on-a-single-device) to measure the model quantization statistics.
```
Let's add a link to it too, it will be easier for readers.

Can you also add the same here please? https://github.com/HabanaAI/optimum-habana-fork/blob/98539470b6c6dbf39145844317d878ddf5482167/examples/text-generation/README.md?plain=1#L600
I had missed that.

--flash_attention_recompute
```

### Saving FP8 Checkpoints in Hugging Face format
After quantizing the model, we can save it to a local path.

**Note:** Before executing the command below, please refer to the "Running FP8 Models on a Single Device" section to measure the model quantization statistics.

Here is an example of how to quantize and save the LLama2-7b model on a single card:
```bash
QUANT_CONFIG=./quantization_config/maxabs_quant_const_scales.json python run_generation.py \

> --model_name_or_path meta-llama/Llama-2-7b-hf \

Reviewer 2: can you change to llama3 model?

> --model_name_or_path meta-llama/Llama-2-7b-hf \

Reviewer 2: please use llama3 -8b instruct as an example

Author: sure

--attn_softmax_bf16 \
--use_hpu_graphs \
--trim_logits \
--use_kv_cache \
--reuse_cache \
--use_flash_attention \
--flash_attention_recompute \
--bf16 \
--batch_size 1 \
--max_new_tokens 128 \
--max_input_tokens 128 \
--limit_hpu_graphs \
--save_quantized_model_with_inc \
--saved_model_path <model_path_on_local_disk>
```

### Loading FP8 Checkpoints saved in Hugging Face format

You can load pre-quantized FP8 models using the `--load_quantized_model_with_inc` argument. The `model_name_or_path` should be the path where the model was saved locally using the previous command in "Saving FP8 Checkpoints in Hugging Face format" section.

Below is an example of how to load a model with FP8 checkpoints on a single card.
Please note that the model name is represented as `<model_path_on_local_disk>`.
```bash
python run_lm_eval.py \
-o acc_load_fp8_model.txt \
--model_name_or_path <model_path_on_local_disk> \
--use_hpu_graphs \
--use_kv_cache \
--trim_logits \
--batch_size 1 \
--bf16 \
--use_flash_attention \
--flash_attention_recompute \
--attn_softmax_bf16 \
--bucket_size=128 \
--bucket_internal \
--load_quantized_model_with_inc
```

> 

Reviewer 2: remove this section and just give examples of llama3

Author: Hi, I don't understand why we need to remove the loading part. Saving and loading is a pair that cannot work without each other.

Author: I think we misled you, what we save is different in huggingface format is different with existing huggingface models. The parameter name in neural magic models is different with INC quantized model.

Reviewer: @xin3he But the command is the same whether loading a model locally or from the hub right? We need to pass `--load_quantized_model_with_inc` and specify the model name (local path or hub name). So I agree with Libin that it would be better to have only one section.

### Loading FP8 Checkpoints from Hugging Face
You can load pre-quantized FP8 models using the `--load_quantized_model_with_inc` argument. The `model_name_or_path` should be a model name from [Neural Magic](https://huggingface.co/collections/neuralmagic/fp8-llms-for-vllm-666742ed2b78b7ac8df13127). For these models, you can use any number of cards to load them.

> 

Reviewer 2: please drop 1 card and keep multi-card example

Author: sure

--flash_attention_recompute
```

### Saving FP8 Checkpoints in Hugging Face format

> After quantizing model, we can save the quantized model into local path.
> Please follow the "Running FP8 models" section to measure model first before running the cmd below.

Reviewer 3: ```suggestion
After quantizing the model, we can save it to a local path.
Please refer to the 'Running FP8 Models' section to measure the model first before executing the command below.
```

> Here is an example to quantize and save LLama2-7b on 1 card:

Reviewer 3: ```suggestion
Here is an example of how to quantize and save the LLama2-7b model on a single card:
```

```bash
QUANT_CONFIG=./quantization_config/maxabs_quant_const_scales.json python run_generation.py \
--model_name_or_path meta-llama/Llama-2-7b-hf \
--attn_softmax_bf16 \
--use_hpu_graphs \
--trim_logits \
--use_kv_cache \
--reuse_cache \
--use_flash_attention \
--flash_attention_recompute \
--bf16 \
--batch_size 1 \
--max_new_tokens 128 \
--max_input_tokens 128 \
--limit_hpu_graphs \
--save_quantized_model_with_inc \
--saved_model_path <model_path_on_local_disk>
```

### Loading FP8 Checkpoints saved in Hugging Face format

> You can load pre-quantized FP8 models with the argument `--load_quantized_model_with_inc`. The `model_name_or_path` is the model path saved on local disk with upper command.

Reviewer 3: ```suggestion
You can load pre-quantized FP8 models using the `--load_quantized_model_with_inc` argument. The `model_name_or_path` should be the path where the model was saved locally using the previous command in "Saving FP8 Checkpoints in Hugging Face format" section.
```

> Below is an example to load a model with FP8 checkpoints on 1 card.
> Please note that model name is denoted as `<model_path_on_local_disk>`

Reviewer 3: ```suggestion
Below is an example of how to load a model with FP8 checkpoints on a single card. 
Please note that the model name is represented as `<model_path_on_local_disk>`.
```

```bash
python run_lm_eval.py \
-o acc_load_fp8_model.txt \
--model_name_or_path <model_path_on_local_disk> \
--use_hpu_graphs \
--use_kv_cache \
--trim_logits \
--batch_size 1 \
--bf16 \
--use_flash_attention \
--flash_attention_recompute \
--attn_softmax_bf16 \
--bucket_size=128 \
--bucket_internal \
--load_quantized_model_with_inc
```

### Loading FP8 Checkpoints from Hugging Face

> You can load pre-quantized FP8 models with the argument `--load_quantized_model_with_inc`. The `model_name_or_path` is a model name from [Neural Magic](https://huggingface.co/collections/neuralmagic/fp8-llms-for-vllm-666742ed2b78b7ac8df13127).
> For models from [Neural Magic](https://huggingface.co/collections/neuralmagic/fp8-llms-for-vllm-666742ed2b78b7ac8df13127), you can use any number of cards to load.

Reviewer 3: ```suggestion
You can load pre-quantized FP8 models using the `--load_quantized_model_with_inc` argument. The `model_name_or_path` should be a model name from [Neural Magic](https://huggingface.co/collections/neuralmagic/fp8-llms-for-vllm-666742ed2b78b7ac8df13127). For these models, you can use any number of cards to load them.
```

> Below is an example to load `neuralmagic/Meta-Llama-3.1-8B-Instruct-FP8` on 1 cards.

Reviewer 3: ```suggestion
Below is an example of how to load `neuralmagic/Meta-Llama-3.1-8B-Instruct-FP8` on a single card.
```

```bash
python run_lm_eval.py \
-o acc_load_fp8_model.txt \
--model_name_or_path neuralmagic/Meta-Llama-3.1-8B-Instruct-FP8 \
--use_hpu_graphs \
--use_kv_cache \
--trim_logits \
--batch_size 1 \
--bf16 \
--use_flash_attention \
--flash_attention_recompute \
--attn_softmax_bf16 \
--bucket_size=128 \
--bucket_internal \
--load_quantized_model_with_inc
```

> Below is an example to load `neuralmagic/Meta-Llama-3.1-70B-Instruct-FP8` on 2 cards.
> 

Reviewer 3: ```suggestion
Below is an example of how to load `neuralmagic/Meta-Llama-3.1-70B-Instruct-FP8` on two cards.
```

```bash
python ../gaudi_spawn.py \
--use_deepspeed --world_size 2 run_lm_eval.py \
-o acc_load_fp8_model.txt \
--model_name_or_path neuralmagic/Meta-Llama-3.1-70B-Instruct-FP8 \
--use_hpu_graphs \
--use_kv_cache \
--trim_logits \
--batch_size 1 \
--bf16 \
--use_flash_attention \
--flash_attention_recompute \
--attn_softmax_bf16 \
--bucket_size=128 \
--bucket_internal \
--load_quantized_model_with_inc

> ```

Reviewer 3: @xin3he the provided command is failing Gaudi2.
 
```sh
[rank0]:   File "/usr/local/lib/python3.10/dist-packages/neural_compressor/torch/algorithms/fp8_quant/save_load.py", line 353, in load
[rank0]:     gathered_state_dict = convert_weight_to_inc(
[rank0]:   File "/usr/local/lib/python3.10/dist-packages/neural_compressor/torch/algorithms/fp8_quant/save_load.py", line 396, in convert_weight_to_inc
[rank0]:     qweight = state_dict[weight_key].t().to(torch.bfloat16).to("hpu")
[rank0]: RuntimeError: [Rank:0] FATAL ERROR :: MODULE:PT_DEVMEM Allocation failed for size::469762048 (448)MB
```

Author: Hi @yafshar  Thanks, you're right.
Here we have two options: one is removing this usage from Readme, the other is adding an function in OH to overwrite incorrect code in INC. What's your opinion?
We didn't find it because we are using nightly environment with this patch. [#48](https://github.com/habana-internal/neural-compressor-fork/pull/48). Also I checked out a new issue from your report and fixed it here [#105](https://github.com/habana-internal/neural-compressor-fork/pull/105)

Reviewer 3: > Hi @yafshar Thanks, you're right. Here we have two options: one is removing this usage from Readme, the other is adding an function in OH to overwrite incorrect code in INC. What's your opinion? We didn't find it because we are using nightly environment with this patch. [#48](https://github.com/habana-internal/neural-compressor-fork/pull/48). Also I checked out a new issue from your report and fixed it here [#105](https://github.com/habana-internal/neural-compressor-fork/pull/105)

@xin3he thanks. If possible, please add a function to overwrite the incorrect code in INC. Additionally, include a #TODO comment to remove this function once INC is released and the issue is resolved.

Author: It's resolved now @yafshar

--saved_model_path <model_path_on_local_disk>
```

> [!NOTE]
> For multi-card usage, the number of cards loaded and used needs to be kept consistent with that when saving.

### Loading FP8 Checkpoints from Hugging Face

> You can load pre-quantized FP8 models using the `--load_quantized_model_with_inc` argument. The `model_name_or_path` should be a model name from [Neural Magic](https://huggingface.co/collections/neuralmagic/fp8-llms-for-vllm-666742ed2b78b7ac8df13127) or a path of FP8 Checkpoints saved in Hugging Face format.

Reviewer: ```suggestion
You can load pre-quantized FP8 models using the `--load_quantized_model_with_inc` argument. The `model_name_or_path` should be a model name from [Neural Magic](https://huggingface.co/collections/neuralmagic/fp8-llms-for-vllm-666742ed2b78b7ac8df13127) or a path to FP8 Checkpoints saved in Hugging Face format.
```
